File: leetcode/No108.convert-sorted-array-to-binary-search-tree.js
/**
 * Difficulty:
 * Easy
 *
 * Desc:
 * Given an array where elements are sorted in ascending order, convert it to a height balanced BST.
 *
 * 为了让二叉树尽量平衡，每次取数组中间值作为树的根节点
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function(nums) {
  if (!nums.length) return null;
  const midIndex = Math.ceil(nums.length / 2) - 1;
  const root = new TreeNode(nums[midIndex]);
  root.left = sortedArrayToBST(nums.slice(0, midIndex));
  root.right = sortedArrayToBST(nums.slice(midIndex + 1));
  return root;
};

File: leetcode/C#/No108.convert-sorted-array-to-binary-search-tree.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given an array where elements are sorted in ascending order, convert it to a height balanced BST.
 * For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1.
 *
 * Example:
 * Given the sorted array: [-10,-3,0,5,9],
 * One possible answer is: [0,-3,9,-10,null,5], which represents the following height balanced BST:
 *      0
 *     / \
 *   -3   9
 *   /   /
 * -10  5
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode SortedArrayToBST(int[] nums) {
        if (nums.Length == 0) return null;
        int index = nums.Length / 2;

        TreeNode node = new TreeNode(nums[index]);
        node.left = SortedArrayToBST(nums[0..index]);
        node.right = SortedArrayToBST(nums[(index + 1)..]);
        return node;
    }
}